feat: migrate cloudflare workers adapter to cloudflare pages

jkhaui opened the pull request:
This PR implements a minimal set of changes so that running waku build with the flag `-- --with-cloudflare` now creates a build specifically for Cloudflare Pages (using their advanced mode strategy). This is the recommended way to deploy web apps on Cloudflare now; worker sites are deprecated.

More discussion and context can be found here: https://github.com/dai-shi/waku/discussions/763

Note: I have tested numerous times that the changes work and can successfully deploy a CF Pages site with no errors. To test this, I used a custom workflow which deploys `examples/10_fs-router` from the monorepo with a compiled version of the waku pnpm workspace. To get this to work I also made the following changes to the example project (these are not committed): 

1. Add `waku.config.ts` with the base path set to `/public/` (explained in the updated docs). 
2. Also needed to add hono as a dependency in the example project, otherwise deploying via Cloudflare would throw an error about a missing hono dependency.

![image](https://github.com/user-attachments/assets/5e1a5539-718e-4a71-b3ad-f70b1d22d843)
![image](https://github.com/user-attachments/assets/1a0adc2d-e69e-4764-9c2d-5d880002134e)


There is a chance that behaviour could be different when consuming Waku as an external npm package. But I can't think of any good ways to replicate this. So I suggest we just test the end-to-end deployment again after merging and releasing this PR

Just let me know if any changes should be made

jkhaui commented:
I wonder what you think about extending `resolveConfig` with a 2nd arg for the deployment platform, like this:

```
// build.ts

const config = await resolveConfig(options.config, options.deploy);
```

and then

```
// config.ts

export async function resolveConfig(config: Config, deploy?: Deploy) {
  const resolvedConfig: ResolvedConfig = {
    basePath: `${deploy !== 'cloudflare' ? '' : '/public'}/`,
    srcDir: 'src',
    distDir: 'dist',
    preserveModuleDirs: ['pages', 'templates', 'routes', 'components'],
    privateDir: 'private',
    rscPath: 'RSC',
    htmlAttrs: '',
    htmlHead: DEFAULT_HTML_HEAD,
    middleware: DEFAULT_MIDDLEWARE,
    ...config,
  };
  if (!resolvedConfig.htmlHead.includes(ADDITIONAL_HTML_HEAD)) {
    resolvedConfig.htmlHead += ADDITIONAL_HTML_HEAD;
  }
  return resolvedConfig;
}
```

Might not be so robust, but for now it would let users deploy to Cloudflare Pages without having to create a `waku.config.ts` with the base path set to `/public` in prod

dai-shi commented:
We will be revisiting our config system later, so let's not make changes for now.

dai-shi commented on packages/waku/src/lib/builder/output-cloudflare.ts:
```diff
@@ -7,21 +7,18 @@ import { DIST_PUBLIC } from './constants.js';
-compatibility_date = "2023-12-06"
-compatibility_flags = [ "nodejs_als" ]
+compatibility_date = "2024-04-05"
+compatibility_flags = ["nodejs_compat"]
```
I was intentionally avoiding nodejs_compat. Do we require it now?

jkhaui replied:
I'm not sure tbh, just assumed it wouldn't work unless nodejs_compat is set. Let me try without it and report back.

Any particular reason you think we should avoid it?

dai-shi replied:
I want to remove nodejs dependency strictly, and cloudflare wrangler gives me errors if some nodejs dependencies remain.

dai-shi commented on packages/waku/src/lib/builder/output-cloudflare.ts:
```diff
@@ -7,21 +7,18 @@ import { DIST_PUBLIC } from './constants.js';
 
-[site]
-bucket = "./${config.distDir}/${DIST_PUBLIC}"
+pages_build_output_dir = "./${config.distDir}/${DIST_PUBLIC}"
```
Hm, I wonder if we can move public files.

```
pages_build_output_dir = "./output"
and
fs.rename("./${config.distDir}/${DIST_PUBLIC}", "./output")
```
something like that.

jkhaui replied:
Hmm. This one I'll probably have to leave to you to decide.

I think I had a similar idea. But I decided not to pursue it further, looked like it was getting very complex because the _worker.js fil then needs to be moved.

In other words, if we implement this strategy, then we must also move _worker.js to `/dist/public/output` as CFP requires _worker.js to exist in the root output directory. There's a good chance other import paths will also break after doing this?

dai-shi replied:
I'm not sure how much it would break others, but I would like to push it for now.
It eliminates the need of `waku.config.ts`, right?

We should revisit it anyway, when we design our build plugin system in the future (not sure when).

jkhaui replied:
In theory it should, I guess the only way is to try it. Let me do this soon and tell you the results :)

Yeah the way I see this solution, it's just to unblock people like me who want to start working immediately with Waku on a platform outside Vercel. I think even just having experimental support for CFP (& other platforms) will bring in more users instead of waiting for a full rewrite of the plugin system.

And users with existing Waku apps should be able to easily upgrade to the new adapter system when it is ready in future

jkhaui replied:
> Hm, I wonder if we can move public files.
> 
> ```
> pages_build_output_dir = "./output"
> and
> fs.rename("./${config.distDir}/${DIST_PUBLIC}", "./output")
> ```
> 
> something like that.

@dai-shi I need to clarify here, do you mean that `dist/public` should be renamed to `dist/output`? And then `pages_build_output_dir = "./dist/output"` Or am I misunderstanding

dai-shi replied:
I'm not sure what's the best, but what I'm suggesting is to restructure (rename) folders in `output-cloudflare.ts` and avoid changing `basePath`.

jkhaui replied:
Got it. Unfortunately the renaming from public to output doesn't look to work

![image](https://github.com/user-attachments/assets/17fe37ce-5a2d-4407-aedd-1ad4df243afc)

I also think it should be left as `public` as this is what [Vite expects](https://vitejs.dev/config/shared-options.html#publicdir)

@rmarscher any suggestions here? Did you encounter any challenges with the `public` directory?

dai-shi commented on packages/waku/src/lib/builder/serve-cloudflare.ts:
```diff
@@ -34,7 +21,6 @@ export default {
   ) {
     if (!serveWaku) {
       serveWaku = runner({ cmd: 'start', loadEntries, env });
-      staticContent = env.__STATIC_CONTENT;
```
I'm happy with this hack going away.

dai-shi reviewed (commented):
first review

jkhaui commented:
Sure thing

I have a question regarding the default value for `preserveModuleDirs`: just curious if there's a reason for not defaulting to a chunked + minified SSR build?

I'm asking because CF has a 1mb max bundle size for workers on their free tier, so every byte in SSR build counts.

Usually I overcome this by setting `vite.build.minify` to use terser, but this doesn't work with Waku - it breaks the build. 

Setting minify to "esbuild" instead of "terser" sometimes works. I've also tried setting `preserveModuleDirs: []` but it doesn't seem to help

jkhaui commented:
Thanks for checking out the PR as well - I'll go through your comments tomorrow 👍

dai-shi commented:
preserveModuleDirs is to make it possible to read files with relative paths, but we might not need to allow it. I don't know why minification fails with or without it.

jkhaui commented on packages/waku/src/lib/builder/build.ts:
```diff
@@ -293,6 +298,24 @@ const buildServerBundle = async (
     throw new Error('Unexpected vite server build output');
   }
+
+  if (serve === 'cloudflare') {
+    // For Cloudflare Pages, create `_routes.json` which prevents calls to
+    // static client assets from invoking the server-side function.
+    // TODO: check for existing `_routes.json` in users' root directory? And
+    // merge this with default staticRoutes provided below
+    const routesJsonFilePath = joinPath(rootDir,
+      config.distDir,
+      '_routes.json');
+    const staticRoutes = {
+      version: 1,
+      include: ['/*'],
+      exclude: ['/public/assets/*']
+    };
+
+    await writeFile(routesJsonFilePath, JSON.stringify(staticRoutes));
+  }
```
I think there's a reason I didn't put it there. Let me check.

And thanks for commenting on this, you made me notice that the public and asset paths should probably be variables instead of hardcoded

jkhaui replied:
I was wrong - we absolutely can (and should) move this to `output-cloudflare.ts`. Good spot!

jkhaui commented on packages/waku/src/lib/builder/build.ts:
```diff
@@ -237,7 +238,11 @@ const buildServerBundle = async (
-              distServeJs: DIST_SERVE_JS,
+              distServeJs: serve !== 'cloudflare'
+                ? DIST_SERVE_JS
+                // Cloudflare Pages advanced mode requires the dist entry file
+                // to be named `_worker.js`
```
Yeah for sure. Astro's CF adapter creates a build output that looks like this:

![image](https://github.com/user-attachments/assets/ee6ce935-cb73-494a-bacf-5a25ee5bc68f)

I don't know how they are using a custom entry path to `_worker/index.js` unless I browse the source code. Because CFP docs say the entry must be called `_worker.js` and be in the output root directory. 

But the point is it means it is possible to customize the `_worker.js` entry file 👍

jkhaui commented on README.md:
````diff
@@ -850,7 +850,20 @@ Adding the `--with-netlify-static` flag to the build script will produce static
+
+```
+export default {
+  basePath: "/public/"
````
Yes I have no idea why this happens either. It could be from CFP, Waku, or Vite's build process. Or even a combination, e.g. CFP might not be able to read `import.meta.env` in some cases.

I definitely see it as temporary. My suggestion is, seeing as Waku doesn't currently have lots of users, we accept an initial hacky workaround. I will then use the published package to see if anything behaves differently and try working on a better solution.

rmarscher commented:
Hi! I've also been working on adjusting the waku cloudflare adapter. I made some progress on the build output. I have it creating the _worker.js as the destination folder, creating a _routes.json and copying the public files to the root of the dist output. I updated the wrangler.toml output to use the [[newer Cloudflare Pages wrangler.toml options](https://developers.cloudflare.com/pages/functions/wrangler-configuration/)](https://developers.cloudflare.com/pages/functions/wrangler-configuration/) (but I’m guessing you already did that here too). I’m just catching up - but I see you found the way some of the other framework adapters handle it.

See https://github.com/dai-shi/waku/compare/main...rmarscher:waku:cloudflare-pages

My changes so far have gotten things working for me when I run `waku build --with-cloudflare` and then run `wrangler pages dev dist`. I haven't added the Cloudflare bindings to the dev server yet. Also, when I try to deploy to cloudflare, I'm getting an error:

```
Failed to publish your Function. Got error: Uncaught Error: No such module "hono".
    imported from "serve.js"
```

I just got that far. I was going to look into why it isn't bundling hono next.

Cloudflare needs a _routes.json file to let it now what are static routes and files that can be cached. I added some code to walk the build public folder and build out the _routes.json. It also can use a generated _headers file to set the cache response headers for certain routes. I didn’t add any custom code for that, but it works if you drop it in your public folder.

I looked at Cloudflare adapters from SvelteKit, Astro and Hono.

- https://github.com/withastro/adapters/blob/main/packages/cloudflare/src/index.ts
- https://github.com/honojs/vite-plugins/blob/main/packages/cloudflare-pages/src/cloudflare-pages.ts
- https://github.com/sveltejs/kit/blob/main/packages/adapter-cloudflare/index.js

They all use the `getPlatformProxy` method from Cloudflare's `wrangler` package to populate the environment bindings in dev mode. https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy

I'm currently trying to figure out a new vite or waku plugin that can handle the wrangler proxy when running `waku dev --with-cloudflare`. Is Vite's configureServer hook the right place to try to inject it? I was going to try that next. https://vitejs.dev/guide/api-plugin#configureserver

The astro adapter seems to be especially thought out. It has deep config options and also does some processing of any redirect files.

jkhaui commented:
> Hi! I've also been working on adjusting the waku cloudflare adapter. I made some progress on the build output. I have it creating the _worker.js as the destination folder, creating a _routes.json and copying the public files to the root of the dist output. I updated the wrangler.toml output to use the [[newer Cloudflare Pages wrangler.toml options](https://developers.cloudflare.com/pages/functions/wrangler-configuration/)](https://developers.cloudflare.com/pages/functions/wrangler-configuration/) (but I’m guessing you already did that here too). I’m just catching up - but I see you found the way some of the other framework adapters handle it.
> 
> 
> 
> See https://github.com/dai-shi/waku/compare/main...rmarscher:waku:cloudflare-pages
> 
> 
> 
> My changes so far have gotten things working for me when I run `waku build --with-cloudflare` and then run `wrangler pages dev dist`. I haven't added the Cloudflare bindings to the dev server yet. Also, when I try to deploy to cloudflare, I'm getting an error:
> 
> 
> 
> ```
> 
> Failed to publish your Function. Got error: Uncaught Error: No such module "hono".
> 
>     imported from "serve.js"
> 
> ```
> 
> 
> 
> I just got that far. I was going to look into why it isn't bundling hono next.
> 
> 
> 
> Cloudflare needs a _routes.json file to let it now what are static routes and files that can be cached. I added some code to walk the build public folder and build out the _routes.json. It also can use a generated _headers file to set the cache response headers for certain routes. I didn’t add any custom code for that, but it works if you drop it in your public folder.
> 
> 
> 
> I looked at Cloudflare adapters from SvelteKit, Astro and Hono.
> 
> 
> 
> - https://github.com/withastro/adapters/blob/main/packages/cloudflare/src/index.ts
> 
> - https://github.com/honojs/vite-plugins/blob/main/packages/cloudflare-pages/src/cloudflare-pages.ts
> 
> - https://github.com/sveltejs/kit/blob/main/packages/adapter-cloudflare/index.js
> 
> 
> 
> They all use the `getPlatformProxy` method from Cloudflare's `wrangler` package to populate the environment bindings in dev mode. https://developers.cloudflare.com/workers/wrangler/api/#getplatformproxy
> 
> 
> 
> I'm currently trying to figure out a new vite or waku plugin that can handle the wrangler proxy when running `waku dev --with-cloudflare`. Is Vite's configureServer hook the right place to try to inject it? I was going to try that next. https://vitejs.dev/guide/api-plugin#configureserver
> 
> 
> 
> The astro adapter seems to be especially thought out. It has deep config options and also does some processing of any redirect files.

That's awesome, great to see someone else here!

I'll read your message more thoroughly in a sec (having lunch atm), but if you want to catchup on our earlier brainstorming, also check out the discussion I linked in the PR description

rmarscher commented:
> That's awesome, great to see someone else here!

:wave:

> I'll read your message more thoroughly in a sec (having lunch atm), but if you want to catchup on our earlier brainstorming, also check out the discussion I linked in the PR description

I got past the hono error. It was being added as an external in the rollup config in the vite-plugin-rsc-serve.ts file if opts.serve was "cloudflare" or "partykit".

That allowed my CF Pages project to deploy... but when I try to open it, I get an error:

```
Error: No such module "ssr/assets/events".
  imported from "ssr/assets/rsc1-3ac5d8483.js"
```

That's coming from this line:
```
import EventEmitter$2, { EventEmitter as EventEmitter$1 } from "events";
```

Maybe that needs a `node:` prefix on the import package name.

I tried with nodejs_als and nodejs_compat and both had the same issue (it did work to build and deploy with nodejs_als so far though...). I also tried the `--no-bundle` option of `wrangler pages deploy` and that didn't seem to make a difference either.

I'll keep plugging away...

rmarscher commented:
Oh actually... it was my client code that was causing the import error. I must have another dependency that has a bundling issue. With just a basic site, it works! https://74bb725c.waku-test.pages.dev/  <- that site loads the message from a json file.

jkhaui commented:
> I got past the hono error. It was being added as an external in the rollup config in the vite-plugin-rsc-serve.ts file if opts.serve was "cloudflare" or "partykit".

Yep good stuff - I was just about to tell you to look there before you posted
 
> That allowed my CF Pages project to deploy... but when I try to open it, I get an error:
> 
> ```
> Error: No such module "ssr/assets/events".
>   imported from "ssr/assets/rsc1-3ac5d8483.js"
> ```
> 
> That's coming from this line:
> 
> ```
> import EventEmitter$2, { EventEmitter as EventEmitter$1 } from "events";
> ```
> 
> Maybe that needs a `node:` prefix on the import package name.
> 
> I tried with nodejs_als and nodejs_compat and both had the same issue (it did work to build and deploy with nodejs_als so far though...). I also tried the `--no-bundle` option of `wrangler pages deploy` and that didn't seem to make a difference either.
> 
> I'll keep plugging away...

Yeah, for the first couple of attempts I made, I also tried to follow a similar implementation as the adapters you listed.

And I also faced a tonne of errors from node imports in transitive dependencies that aren't prefixed with `node:`. The challenge (if I understand correctly) is that you can only fix these errors by updating the source code of every incompatible dependency and add the `node:` prefix. Though I'm also unsure why these errors occur in some implementations of Waku + CFP but not others (i.e. like the way I've done this PR). 

How are you deploying to CFP? Are you using a connected git repo? IIRC this also affected the build process behaviour for me

Ultimately, I decided it's not worth the effort atm to figure out all these quirks if the minimal change of renaming `serve.js` to `_worker.js` (along with a couple of other minor changes) appears to be enough to get basic CFP working. I also didn't even look into setting up a platform proxy with Wrangler - as @dai-shi has mentioned, Waku's adapter system will be revamped in future. When this happens it could be a good time to add any dev environment features?

The stuff you've done so far looks great. though. Seeing as our implementations are quite different, I suggest we first aim to merge in this PR with its minimal changes and verify it works when Waku is consumed as an external package. From here, we can then cherry-pick any enhancements from the work you've been doing. E.g. support for CF bindings looks a good candidate

What does everyone think?

rmarscher commented:
> The stuff you've done so far looks great. though. Seeing as our implementations are quite different, I suggest we first aim to merge in this PR with its minimal changes and verify it works when Waku is consumed as an external package. From here, we can then cherry-pick any enhancements from the work you've been doing. E.g. support for CF bindings looks a good candidate

I submitted a PR to make it a bit easier to review and try the changes I made. I'm heading to bed, but feel free to handle this in any way -- merge this PR and then I can potentially rebase and resubmit or however everyone wants to handle it.

> **edit:** just saw your last message @rmarscher , so you got it working? Nice! Are you able to share what your `dist` directory structure looks like?

Yup. It looks like this. Cheers!

<img width="236" alt="Screenshot 2024-07-17 at 2 39 44 AM" src="https://github.com/user-attachments/assets/667f4c7d-2e84-494f-9d90-a9bdd7858f63">

jkhaui commented:
Closing this so we can consolidate efforts behind [this PR](https://github.com/dai-shi/waku/pull/795). Can always reopen if we change our mind for whatever reason.
